leader/leader.go

```diff
@@ -124,229 +121,239 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
 }
 
 // AddWatch starts watching the key for changes and sending them
-// to the valuesC, the watch is stopped
-func (l *Client) AddWatch(key string, retry time.Duration, valuesC chan string) {
-	prefix := fmt.Sprintf("AddWatch(key=%v)", key)
+// to the valuesC.
+// The watch can be stopped either with the specified context or Client.Close
+func (l *Client) AddWatch(ctx context.Context, key string, retry time.Duration, valuesC chan string) {
 	api := client.NewKeysAPI(l.client)
-
-	go func() {
+	entry := log.WithFields(log.Fields{"watch": key})
+	renewCtx, cancel := context.WithCancel(ctx)
+	renew := func() (client.Watcher, *client.Response, error) {
+		b := NewUnlimitedExponentialBackOff()
+		b.InitialInterval = retry
 		var watcher client.Watcher
 		var resp *client.Response
-		var err error
-
-		ctx, closer := context.WithCancel(context.WithValue(context.Background(), "prefix", prefix))
-		go func() {
-			<-l.closeC
-			closer()
-		}()
-
-		backoff := &FreebieExponentialBackOff{
-			InitialInterval:     500 * time.Millisecond,
-			RandomizationFactor: 0.5,
-			Multiplier:          1.5,
-			MaxInterval:         10 * time.Second,
-			MaxElapsedTime:      30 * time.Second,
+		err := backoff.Retry(func() (err error) {
+			watcher, resp, err = l.recreateWatchAtLatestIndex(renewCtx, api, key, entry)
+			return trace.Wrap(err)
+		}, backoff.WithContext(b, renewCtx))
+		if err != nil {
+			return nil, nil, trace.Wrap(err)
 		}
-		ticker := ebackoff.NewTicker(backoff)
+		return watcher, resp, nil
+	}
 
-		watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
-		if err != nil {
+	go func() {
+		defer cancel()
+		select {
+		case <-ctx.Done():
+			return
+		case <-l.closeC:
 			return
 		}
+	}()
 
-		// make sure we always send the first actual value
-		if resp != nil && resp.Node != nil {
+	go l.watcher(ctx, renew, valuesC, entry)
+}
+
+func (l *Client) watcher(ctx context.Context, renew renewFunc, valuesC chan string, entry *log.Entry) {
```

Comment:
this is a very complex logic, and became even more complex. Why all these changes?